# Исследование клиентской воронки покупки авиабилетов

## 1. Подготовка и обзор данных

In [1]:
import pandas as pd
import numpy as np
import plotly.express as px
import plotly.graph_objects as go

In [2]:
df = pd.read_excel('dano_dataset_avia_hashed_corrected_s.xlsx')

In [3]:
df.head()

,session_start_dt,device_platform,client_start_year,client_id,session_id,session_source,session_channel,screen_nm,screen_dttm,client_avia_user_start_year,...,client_gmv,session_purchase_price,session_purchase_tickets_num,session_purchase_tickets_sourse,session_purchase_status,session_purchase_tickets_airline,session_purchase_tickets_departure,session_purchase_tickets_arrival,session_purchase_tickets_luggage_flg,session_purchase_tickets_class_of_service
0,2025-08-11,android,2022.0,1ced90f0cbf7673a49e1f06035a20c2d31bd2876e768f2...,168e64115c5eb71efa3f2ffb59a38a04f182540de70312...,Контент-каналы,Сторис,Оффер,2025-08-11 13:05:44.638,2025,...,27427.476667,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2025-04-10,ios,2016.0,707572ff842ece6fd9c6ab59c5f7d7309de18ae153995b...,a1e18971d51f0e7699396daa17b8373dd2bddec4ffa2da...,Неизвестно,Неизвестно,Оффер,2025-04-10 23:55:10.936,2024,...,72363.028889,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,2025-06-05,ios,2025.0,514278f8446c5920e20faddc6aa5b03cc8fa784aa9be9a...,f29b5c0c7391fb9aee4bb47d9dab77d1161e1eec5b108f...,SEO,Bing,Выдача,2025-06-05 10:23:44.627,2025,...,6645.640000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,2025-02-18,android,2022.0,d42de563a20cd93b6218c5eab57aac2f62e8a6ffaec3d1...,9abb7f52910e38152f4e261157fd4435534dbf7273f080...,SEO,Яндекс,Поиск,2025-02-18 15:21:08.491,2024,...,3543.480000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,2025-09-01,ios,2022.0,fb3597c695c46ebe63eb6f8ff9e89b7cd0fc74a6c309fe...,a6a3488345af3d422b8bf2b11c14f1aa75c4829c606179...,Неизвестно,Неизвестно,Выдача,2025-09-01 11:35:23.627,2024,...,22816.832000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [4]:
cols = [
    'client_start_year',
    'client_avia_user_start_year',
    'client_segment',
    'client_gmv',
    'client_avia_orders',
    'session_id',
    'session_start_dt',
    'device_platform',
    'session_source',
    'session_channel',
    'screen_nm',
    'screen_dttm'
]

df[cols].head()

,client_start_year,client_avia_user_start_year,client_segment,client_gmv,client_avia_orders,session_id,session_start_dt,device_platform,session_source,session_channel,screen_nm,screen_dttm
0,2022.0,2025,2,27427.476667,12,168e64115c5eb71efa3f2ffb59a38a04f182540de70312...,2025-08-11,android,Контент-каналы,Сторис,Оффер,2025-08-11 13:05:44.638
1,2016.0,2024,2,72363.028889,36,a1e18971d51f0e7699396daa17b8373dd2bddec4ffa2da...,2025-04-10,ios,Неизвестно,Неизвестно,Оффер,2025-04-10 23:55:10.936
2,2025.0,2025,5,6645.640000,1,f29b5c0c7391fb9aee4bb47d9dab77d1161e1eec5b108f...,2025-06-05,ios,SEO,Bing,Выдача,2025-06-05 10:23:44.627
3,2022.0,2024,4,3543.480000,1,9abb7f52910e38152f4e261157fd4435534dbf7273f080...,2025-02-18,android,SEO,Яндекс,Поиск,2025-02-18 15:21:08.491
4,2022.0,2024,2,22816.832000,125,a6a3488345af3d422b8bf2b11c14f1aa75c4829c606179...,2025-09-01,ios,Неизвестно,Неизвестно,Выдача,2025-09-01 11:35:23.627


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244345 entries, 0 to 244344
Data columns (total 22 columns):
 #   Column                                     Non-Null Count   Dtype  
---  ------                                     --------------   -----  
 0   session_start_dt                           244345 non-null  object 
 1   device_platform                            244134 non-null  object 
 2   client_start_year                          244300 non-null  float64
 3   client_id                                  244345 non-null  object 
 4   session_id                                 244345 non-null  object 
 5   session_source                             244345 non-null  object 
 6   session_channel                            244345 non-null  object 
 7   screen_nm                                  244345 non-null  object 
 8   screen_dttm                                244345 non-null  object 
 9   client_avia_user_start_year                244345 non-null  int64  
 10  client_s

In [6]:
df.isna().mean()


session_start_dt                             0.000000
device_platform                              0.000864
client_start_year                            0.000184
client_id                                    0.000000
session_id                                   0.000000
session_source                               0.000000
session_channel                              0.000000
screen_nm                                    0.000000
screen_dttm                                  0.000000
client_avia_user_start_year                  0.000000
client_segment                               0.000000
client_avia_orders                           0.000000
client_gmv                                   0.000000
session_purchase_price                       0.988586
session_purchase_tickets_num                 0.988586
session_purchase_tickets_sourse              0.989269
session_purchase_status                      0.989269
session_purchase_tickets_airline             0.989269
session_purchase_tickets_dep

In [7]:
df.isna().sum()

session_start_dt                                  0
device_platform                                 211
client_start_year                                45
client_id                                         0
session_id                                        0
session_source                                    0
session_channel                                   0
screen_nm                                         0
screen_dttm                                       0
client_avia_user_start_year                       0
client_segment                                    0
client_avia_orders                                0
client_gmv                                        0
session_purchase_price                       241556
session_purchase_tickets_num                 241556
session_purchase_tickets_sourse              241723
session_purchase_status                      241723
session_purchase_tickets_airline             241723
session_purchase_tickets_departure           241723
session_purc

In [8]:
df = df.dropna(subset=['device_platform', 'client_start_year'])


In [9]:
df[['device_platform', 'client_start_year']].isna().sum()


device_platform      0
client_start_year    0
dtype: int64

Пропуски в признаках device_platform и client_start_year составили менее 0.001% от общего числа наблюдений. В связи с незначительной долей пропусков было принято решение удалить соответствующие строки. Такой подход позволил сохранить качество данных без заметной потери информации.

In [10]:
df['session_start_dt'] = pd.to_datetime(df['session_start_dt'])
df['screen_dttm'] = pd.to_datetime(df['screen_dttm'])
df['client_start_year'] = df['client_start_year'].astype(int)
df['month'] = df['session_start_dt'].dt.to_period('M').dt.to_timestamp()

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 244089 entries, 0 to 244344
Data columns (total 23 columns):
 #   Column                                     Non-Null Count   Dtype         
---  ------                                     --------------   -----         
 0   session_start_dt                           244089 non-null  datetime64[ns]
 1   device_platform                            244089 non-null  object        
 2   client_start_year                          244089 non-null  int64         
 3   client_id                                  244089 non-null  object        
 4   session_id                                 244089 non-null  object        
 5   session_source                             244089 non-null  object        
 6   session_channel                            244089 non-null  object        
 7   screen_nm                                  244089 non-null  object        
 8   screen_dttm                                244089 non-null  datetime64[ns]
 9   client_av

In [12]:
df['screen_nm'].value_counts().head(30)

screen_nm
Выдача           110486
Поиск             59392
Оффер             53162
Чекаут            15832
Покупка бэк        2787
Покупка фронт      2430
Name: count, dtype: int64

## 2. Общий анализ воронки

In [13]:
stage_order = [
    'Поиск',
    'Выдача',
    'Оффер',
    'Чекаут',
    'Успешная покупка'
]

stage_values = [
    df.loc[df['screen_nm'] == 'Поиск', 'session_id'].nunique(),
    df.loc[df['screen_nm'] == 'Выдача', 'session_id'].nunique(),
    df.loc[df['screen_nm'] == 'Оффер', 'session_id'].nunique(),
    df.loc[df['screen_nm'] == 'Чекаут', 'session_id'].nunique(),
    df.loc[df['session_purchase_status'] == 'Успешные', 'session_id'].nunique()
]

funnel_df = pd.DataFrame({
    'stage': stage_order,
    'sessions': stage_values
})

fig = go.Figure(go.Funnel(
    y=funnel_df['stage'],
    x=funnel_df['sessions'],
    textinfo='value+percent initial+percent previous'
))

fig.update_layout(
    title='Воронка покупки авиабилетов',
    xaxis_title='Количество уникальных сессий',
    yaxis_title='Этап'
)

fig.show()

На этапе "Поиск" наблюдается 37 718 уникальных сессий, что соответствует 100% от общего количества. На следующем этапе "Выдача" количество сессий сокращается до 34 286, что составляет 91% от начального этапа. Далее, на этапе "Оффер", количество сессий уменьшается до 17 703, что составляет 46.9% от начального этапа и 51.6% от предыдущего этапа. На этапе "Чекаут" количество сессий сокращается до 6466, что составляет 17.1% от начального этапа и 36.5% от предыдущего этапа. Наконец, на этапе "Покупка" количество сессий составляет 2362, что соответствует 6.3% от начального этапа и 36.5% от предыдущего этапа. 

### Таблица потерь по шагам воронки

In [14]:
funnel_table = funnel_df.copy()

funnel_table['drop_from_prev'] = funnel_table['sessions'].shift(1) - funnel_table['sessions']
funnel_table.loc[0, 'drop_from_prev'] = 0

funnel_table['cr_from_prev'] = funnel_table['sessions'] / funnel_table['sessions'].shift(1)
funnel_table.loc[0, 'cr_from_prev'] = 1

funnel_table['cr_from_start'] = funnel_table['sessions'] / funnel_table['sessions'].iloc[0]

funnel_table['drop_pct_from_prev'] = 1 - funnel_table['cr_from_prev']
funnel_table.loc[0, 'drop_pct_from_prev'] = 0

percent_cols = ['cr_from_prev', 'cr_from_start', 'drop_pct_from_prev']
count_cols = ['sessions', 'drop_from_prev']

funnel_table[percent_cols] = (funnel_table[percent_cols] * 100).round(2)
funnel_table[count_cols] = funnel_table[count_cols].astype(int)

funnel_table

,stage,sessions,drop_from_prev,cr_from_prev,cr_from_start,drop_pct_from_prev
0,Поиск,37718,0,100.00,100.00,0.00
1,Выдача,34286,3432,90.90,90.90,9.10
2,Оффер,17703,16583,51.63,46.94,48.37
3,Чекаут,6466,11237,36.52,17.14,63.48
4,Успешная покупка,2362,4104,36.53,6.26,63.47


По таблице потерь видно, что основная абсолютная потеря пользователей происходит на переходе `Выдача → Оффер`: на этом шаге воронка теряет 16.6 тыс. сессий, или 48.4% пользователей предыдущего этапа. Это делает экран выдачи одной из ключевых точек для продуктовой оптимизации.

При этом наиболее низкая относительная конверсия наблюдается в нижней части воронки: переходы `Оффер → Чекаут` и `Чекаут → Успешная покупка` имеют конверсию около 36.5%, то есть на каждом из этих шагов теряется почти две трети пользователей. В результате итоговая конверсия от поиска до успешной покупки составляет 6.26%.

## 3. Углубление в проблемный шаг Выдача -> Оффер

In [15]:
screen_times = (
    df[df['screen_nm'].isin(['Выдача', 'Оффер'])]
    .sort_values(['session_id', 'screen_dttm'])
    .groupby(['session_id', 'screen_nm'], as_index=False)['screen_dttm']
    .first()
    .pivot(index='session_id', columns='screen_nm', values='screen_dttm')
    .reset_index()
)
screen_times = screen_times[screen_times['Выдача'].notna()].copy()
screen_times['reached_offer'] = (
        screen_times['Оффер'].notna() &
        (screen_times['Оффер'] >= screen_times['Выдача'])
)

last_screen_time = (
    df.groupby('session_id', as_index=False)['screen_dttm']
    .max()
    .rename(columns={'screen_dttm': 'last_screen_dttm'})
)

screen_times = screen_times.merge(last_screen_time, on='session_id', how='left')

screen_times['time_on_results_sec'] = np.where(
    screen_times['reached_offer'],
    (screen_times['Оффер'] - screen_times['Выдача']).dt.total_seconds(),
    (screen_times['last_screen_dttm'] - screen_times['Выдача']).dt.total_seconds()
)

screen_times = screen_times[screen_times['time_on_results_sec'] > 0].copy()

bins = [-np.inf, 10, 30, 60, 120, 300, 600, 1200, 3600, np.inf]
labels = ['<=10с', '10–30с', '30–60с', '1–2м', '2–5м', '5–10м', '10–20м', '20–60м', '60м+']

screen_times['time_bucket'] = pd.cut(
    screen_times['time_on_results_sec'],
    bins=bins,
    labels=labels,
    right=True
)
bucket_stats = (
    screen_times.groupby('time_bucket', observed=False).agg(sessions_cnt=('session_id', 'nunique'), reached_offer_share=('reached_offer', 'mean')).reset_index())

bucket_stats

,time_bucket,sessions_cnt,reached_offer_share
0,<=10с,1948,0.956879
1,10–30с,5786,0.841860
2,30–60с,5107,0.675935
3,1–2м,4425,0.621017
4,2–5м,3488,0.563647
5,5–10м,1363,0.577403
6,10–20м,934,0.513919
7,20–60м,1193,0.499581
8,60м+,1973,0.466802


In [16]:
fig = px.bar(
    bucket_stats,
    x='time_bucket',
    y='reached_offer_share',
    text='sessions_cnt',
    title='Переход Выдача -> Оффер в зависимости от времени на Выдаче',
    labels={
        'time_bucket': 'Время на Выдаче',
        'reached_offer_share': 'Доля сессий, дошедших до Оффера'
    }
)

fig.update_yaxes(range=[0, 1.08], tickformat='.0%')
fig.update_traces(
    texttemplate='n=%{text}',
    textposition='outside'
)

fig.update_layout(
    xaxis_title='Время на Выдаче',
    yaxis_title='Доля сессий, дошедших до Оффера'
)

fig.show()

По графику видно, что пользователи, которые быстро переходят с выдачи к офферу, чаще доходят до следующего шага. А если пользователь проводит на выдаче много времени, вероятность перехода в оффер снижается.

Это может говорить о том, что часть пользователей быстро находит подходящий вариант и сразу открывает его. А те, кто долго остаётся на выдаче, скорее сталкиваются со сложностью выбора или не находят достаточно подходящее предложение.

Значит, экран выдачи можно считать одной из важных точек для улучшения продукта.

### Время перехода от чекаута к успешной покупке
Посмотрим, сколько времени в среднем пользователи тратят на переход от этапа `Чекаут` к успешной покупке.

In [17]:
checkout_purchase_times = (
    df[df['screen_nm'] == 'Чекаут']
    .sort_values(['session_id', 'screen_dttm'])
    .groupby('session_id', as_index=False)['screen_dttm']
    .first()
    .rename(columns={'screen_dttm': 'checkout_time'})
)

purchase_times = (
    df[df['session_purchase_status'] == 'Успешные']
    .sort_values(['session_id', 'screen_dttm'])
    .groupby('session_id', as_index=False)['screen_dttm']
    .max()
    .rename(columns={'screen_dttm': 'purchase_time'})
)

checkout_purchase_times = checkout_purchase_times.merge(
    purchase_times,
    on='session_id',
    how='inner'
)

checkout_purchase_times = checkout_purchase_times[
    checkout_purchase_times['purchase_time'] >= checkout_purchase_times['checkout_time']
].copy()

checkout_purchase_times['checkout_to_purchase_sec'] = (
    checkout_purchase_times['purchase_time'] - checkout_purchase_times['checkout_time']
).dt.total_seconds()

checkout_purchase_summary = pd.DataFrame({
    'metric': ['mean_sec', 'median_sec', 'mean_min', 'median_min'],
    'value': [
        checkout_purchase_times['checkout_to_purchase_sec'].mean(),
        checkout_purchase_times['checkout_to_purchase_sec'].median(),
        checkout_purchase_times['checkout_to_purchase_sec'].mean() / 60,
        checkout_purchase_times['checkout_to_purchase_sec'].median() / 60,
    ]
})

checkout_purchase_summary['value'] = checkout_purchase_summary['value'].round(2)
checkout_purchase_summary

,metric,value
0,mean_sec,2124.14
1,median_sec,402.57
2,mean_min,35.40
3,median_min,6.71


По таблице видно, что пользователи, которые дошли до чекаута и завершили покупку, обычно делают это довольно быстро: медианное время от чекаута до покупки составляет около 6.7 минуты.

При этом среднее время заметно выше — около 35 минут. Это значит, что в данных есть часть длинных сессий, которые увеличивают среднее значение. Поэтому медиана здесь лучше показывает типичное поведение пользователя.

В целом можно сказать, что успешная покупка после перехода в чекаут чаще всего происходит в течение нескольких минут, но у части пользователей процесс заметно затягивается.

## 4. Динамика воронки по месяцам

In [18]:
users_by_month = (
    df.groupby('month')['client_id']
    .nunique()
    .reset_index()
    .rename(columns={'client_id': 'users_cnt'})
)
fig = px.bar(
    users_by_month,
    x='month',
    y='users_cnt',
    title='Количество уникальных пользователей по месяцам',
    labels={
        'month': 'Месяц',
        'users_cnt': 'Количество пользователей'
    }
)
fig.update_traces(textposition='outside')
fig.show()

In [19]:
sessions_by_month = (
    df.groupby('month')['session_id']
    .nunique()
    .reset_index()
    .rename(columns={'session_id': 'sessions_cnt'})
)
fig = px.bar(
    sessions_by_month,
    x='month',
    y='sessions_cnt',
    title='Количество сессий по месяцам',
    labels={
        'month': 'Месяц',
        'sessions_cnt': 'Количество сессий'
    }
)
fig.update_traces(textposition='outside')
fig.show()

Количество пользователей и сессий по месяцам в целом стабильное, посмотрим количество покупок и конверсию в покупку по месяцам, чтобы выявить возможные аномалии.

In [20]:
purchases_by_month = (
    df.loc[df['session_purchase_status'] == 'Успешные']
    .groupby('month')['session_id']
    .nunique()
    .reset_index()
    .rename(columns={'session_id': 'purchases_cnt'})
)

conversion_by_month = sessions_by_month.merge(
    purchases_by_month,
    on='month',
    how='left'
)

conversion_by_month['purchases_cnt'] = conversion_by_month['purchases_cnt'].fillna(0)
conversion_by_month['conversion_to_purchase'] = (
    conversion_by_month['purchases_cnt'] / conversion_by_month['sessions_cnt']
)

conversion_by_month

,month,sessions_cnt,purchases_cnt,conversion_to_purchase
0,2025-01-01,4783,394,0.082375
1,2025-02-01,4002,263,0.065717
2,2025-03-01,4362,294,0.067400
3,2025-04-01,4471,280,0.062626
4,2025-05-01,4650,282,0.060645
5,2025-06-01,4575,264,0.057705
6,2025-07-01,4061,82,0.020192
7,2025-08-01,4388,293,0.066773
8,2025-09-01,3231,210,0.064995


In [21]:
monthly_metrics = sessions_by_month.merge(
    purchases_by_month,
    on='month',
    how='left'
)

monthly_metrics['purchases_cnt'] = monthly_metrics['purchases_cnt'].fillna(0)
monthly_metrics['conversion_to_purchase'] = (
    monthly_metrics['purchases_cnt'] / monthly_metrics['sessions_cnt']
)
monthly_metrics['conversion_pct'] = (
    monthly_metrics['conversion_to_purchase'] * 100
).round(2)

fig = px.bar(
    monthly_metrics,
    x='month',
    y='purchases_cnt',
    title='Количество покупок по месяцам',
    labels={
        'month': 'Месяц',
        'purchases_cnt': 'Количество покупок'
    },
    text='purchases_cnt'
)

fig.update_traces(textposition='outside')
fig.show()

In [22]:
fig = px.line(
    monthly_metrics,
    x='month',
    y='conversion_to_purchase',
    markers=True,
    title='Конверсия в покупку по месяцам',
    labels={
        'month': 'Месяц',
        'conversion_to_purchase': 'Конверсия в покупку'
    }
)

fig.update_yaxes(tickformat='.1%')
fig.show()

Анализ динамики покупок и конверсии по месяцам показал, что в большинстве месяцев метрики оставались относительно стабильными. Однако в июле наблюдается резкое снижение как количества покупок, так и конверсии в покупку: число покупок сократилось до 82, а конверсия — до 2.0%. Поскольку одновременно снизились и абсолютный объем покупок, и эффективность перевода сессий в покупку, июль можно считать аномальным месяцем.

In [23]:
def build_funnel(data, label):
    result = pd.DataFrame({
        'stage': stage_order,
        'sessions': [
            data.loc[data['screen_nm'] == 'Поиск', 'session_id'].nunique(),
            data.loc[data['screen_nm'] == 'Выдача', 'session_id'].nunique(),
            data.loc[data['screen_nm'] == 'Оффер', 'session_id'].nunique(),
            data.loc[data['screen_nm'] == 'Чекаут', 'session_id'].nunique(),
            data.loc[data['session_purchase_status'] == 'Успешные', 'session_id'].nunique()
        ]
    })
    result['group'] = label
    result['cr_from_start'] = result['sessions'] / result['sessions'].iloc[0]
    result['cr_from_prev'] = result['sessions'] / result['sessions'].shift(1)
    result.loc[0, 'cr_from_prev'] = 1
    return result

july = build_funnel(df[df['month'] == pd.Timestamp('2025-07-01')], 'Июль')
august = build_funnel(df[df['month'] == pd.Timestamp('2025-08-01')], 'Август')
other = build_funnel(
    df[~df['month'].isin([pd.Timestamp('2025-07-01'), pd.Timestamp('2025-08-01')])],
    'Остальные месяцы'
)

funnel_compare = pd.concat([july, august, other], ignore_index=True)

fig = px.bar(
    funnel_compare,
    x='stage',
    y='cr_from_start',
    color='group',
    barmode='group',
    title='Сравнение конверсии по шагам: июль, август и остальные месяцы',
    labels={
        'stage': 'Этап воронки',
        'cr_from_start': 'Конверсия от первого шага',
        'group': 'Период'
    },
    text='cr_from_start'
)

fig.update_yaxes(tickformat='.1%')
fig.update_traces(texttemplate='%{text:.1%}', textposition='outside')
fig.show()

График показывает, что проблема июля сосредоточена в нижней части воронки. Верхний этап (поиск -> выдача) просел незначительно, однако начиная с оффера июль заметно отстаёт от августа и остальных месяцев. Особенно сильная разница видна на этапах чекаута и успешной покупки: итоговая конверсия июля составляет 2.1% против 6.9% в августе и 6.7% в остальные месяцы. Это указывает на локальную проблему июля в процессе оформления или покупки, а не на снижение объёма или качества входящего трафика.

## 5. Сегментация конверсии по каналам привлечения

### Конверсия в покупку по session_channel

In [24]:
channel_sessions = (
    df.groupby('session_channel')['session_id']
    .nunique()
    .reset_index()
    .rename(columns={'session_id': 'sessions_cnt'})
)

channel_purchases = (
    df.loc[df['session_purchase_status'] == 'Успешные']
    .groupby('session_channel')['session_id']
    .nunique()
    .reset_index()
    .rename(columns={'session_id': 'purchases_cnt'})
)

channel_conversion = channel_sessions.merge(
    channel_purchases,
    on='session_channel',
    how='left'
)

channel_conversion['purchases_cnt'] = channel_conversion['purchases_cnt'].fillna(0)
channel_conversion['conversion_to_purchase'] = (
    channel_conversion['purchases_cnt'] / channel_conversion['sessions_cnt']
)
channel_conversion = channel_conversion.sort_values('conversion_to_purchase', ascending=False)

channel_conversion

,session_channel,sessions_cnt,purchases_cnt,conversion_to_purchase
3,banner,1912,137,0.071653
1,Google,3886,257,0.066135
4,email,1910,126,0.065969
2,UTM-метки Выгоды,1924,118,0.061331
5,Неизвестно,19241,1179,0.061275
0,Bing,1903,113,0.059380
6,Сторис,3806,219,0.057541
7,Яндекс,3941,213,0.054047


In [25]:
fig = px.bar(
    channel_conversion,
    x='session_channel',
    y='conversion_to_purchase',
    title='Конверсия в покупку по session_channel',
    labels={
        'session_channel': 'Канал',
        'conversion_to_purchase': 'Конверсия в покупку'
    },
    text='conversion_to_purchase'
)

fig.update_yaxes(tickformat='.1%', range=[0, 0.08])

fig.update_traces(
    texttemplate='%{text:.1%}',
    textposition='outside',
    textfont_size=14,
    width=0.7
)

fig.update_layout(
    height=650,
    xaxis_tickangle=-35,
    uniformtext_minsize=10,
    uniformtext_mode='hide',
    margin=dict(t=120, b=120, l=80, r=40)
)

fig.show()

Конверсия в покупку по session_channel различается умеренно - от 5.4% до 7.2%. Наиболее высокий уровень конверсии наблюдается у banner, Google и email, тогда как наиболее низкий - у Яндекса и Сторис. При этом каналы существенно различаются по объёму трафика. На текущем этапе можно предположить, что отдельные каналы различаются по качеству привлечения, однако для уверенного вывода необходимо дополнительно проверить статистическую значимость различий и отдельно проанализировать поведение каналов в июле.

## 6. Дополнительный контекст по покупкам

In [26]:
purchase_price_by_month = (
    df[df['session_purchase_status'] == 'Успешные']
    .groupby('month')
    .agg(
        purchases_cnt=('session_id', 'nunique'),
        avg_purchase_price=('session_purchase_price', 'mean'),
        median_purchase_price=('session_purchase_price', 'median')
    )
    .reset_index()
)

purchase_price_by_month

,month,purchases_cnt,avg_purchase_price,median_purchase_price
0,2025-01-01,394,14490.024286,9842.340
1,2025-02-01,263,13810.117085,8649.830
2,2025-03-01,294,14153.207151,10141.205
3,2025-04-01,280,13592.214188,10042.880
4,2025-05-01,282,14945.061497,9200.960
5,2025-06-01,264,15003.014877,10810.540
6,2025-07-01,82,16754.383667,10849.350
7,2025-08-01,293,17365.344455,12142.740
8,2025-09-01,210,17100.228106,12256.970


In [27]:
fig = px.line(
    purchase_price_by_month,
    x='month',
    y='avg_purchase_price',
    markers=True,
    title='Средняя цена успешной покупки по месяцам',
    labels={
        'month': 'Месяц',
        'avg_purchase_price': 'Средняя цена покупки'
    }
)

fig.show()

### Динамика средней цены покупки по месяцам

Анализ показал, что начиная с июля средняя и медианная цена успешной покупки заметно выросли относительно первой половины года. В июле средняя цена составила 16.8 тыс., тогда как в январе–июне она находилась в диапазоне примерно 13.6–15.0 тыс. При этом в августе и сентябре уровень цен сохранился и даже оказался выше июльского.

Это означает, что июльская просадка конверсии не может быть объяснена только ростом цены. Если бы основной причиной было исключительно удорожание покупки, то снижение конверсии должно было бы сохраняться и в августе–сентябре, когда средняя цена оставалась высокой. Однако после июля конверсия восстановилась.

Следовательно, рост среднего чека мог быть сопутствующим фактором, но не выглядит основной причиной июльской аномалии. Для дальнейшего анализа необходимо проверить не только уровень цены, но и структуру покупок: число билетов, направления, класс обслуживания, продавцов и другие характеристики офферов.

In [28]:
tickets_by_month = (
    df[df['session_purchase_status'] == 'Успешные']
    .groupby('month')
    .agg(
        purchases_cnt=('session_id', 'nunique'),
        avg_tickets_num=('session_purchase_tickets_num', 'mean'),
        median_tickets_num=('session_purchase_tickets_num', 'median')
    )
    .reset_index()
)

tickets_by_month

,month,purchases_cnt,avg_tickets_num,median_tickets_num
0,2025-01-01,394,1.477752,1.0
1,2025-02-01,263,1.484746,1.0
2,2025-03-01,294,1.523256,1.0
3,2025-04-01,280,1.464286,1.0
4,2025-05-01,282,1.611465,1.0
5,2025-06-01,264,1.557895,1.0
6,2025-07-01,82,1.677778,1.0
7,2025-08-01,293,1.784848,1.0
8,2025-09-01,210,1.748899,1.0


In [29]:
fig = px.line(
    tickets_by_month,
    x='month',
    y='avg_tickets_num',
    markers=True,
    title='Среднее количество билетов в успешной покупке по месяцам',
    labels={
        'month': 'Месяц',
        'avg_tickets_num': 'Среднее количество билетов'
    }
)

fig.show()

### Динамика количества билетов в покупке по месяцам

Анализ показал, что начиная с мая–июля в успешных покупках растёт среднее количество билетов. Если в январе–апреле среднее значение находилось в диапазоне 1.46–1.52, то в июле оно достигло 1.68, а в августе–сентябре — 1.75–1.78. При этом медианное значение на всём периоде остаётся равным 1, что указывает на сохранение доминирующего сценария покупки одного билета; рост среднего значения, вероятно, связан с увеличением доли заказов с несколькими билетами.

Это означает, что рост среднего чека с июля частично может объясняться не только изменением уровня цен, но и увеличением числа билетов в одной покупке. Однако сам по себе этот фактор не объясняет июльскую просадку конверсии: в августе и сентябре среднее количество билетов остаётся ещё выше, тогда как конверсия в покупку восстанавливается.

Следовательно, изменение размера заказа могло сопровождать июльскую аномалию, но не выглядит её основной причиной.

## 7. Сводка ключевых метрик

In [30]:
monthly_summary = (
    sessions_by_month
    .merge(users_by_month, on='month', how='left')
    .merge(purchases_by_month, on='month', how='left')
    .merge(
        purchase_price_by_month[['month', 'avg_purchase_price', 'median_purchase_price']],
        on='month',
        how='left'
    )
    .merge(
        tickets_by_month[['month', 'avg_tickets_num', 'median_tickets_num']],
        on='month',
        how='left'
    )
)

monthly_summary['purchases_cnt'] = monthly_summary['purchases_cnt'].fillna(0)
monthly_summary['conversion_to_purchase'] = (
    monthly_summary['purchases_cnt'] / monthly_summary['sessions_cnt']
)

monthly_summary['conversion_pct'] = (
    monthly_summary['conversion_to_purchase'] * 100
).round(2)

monthly_summary = monthly_summary[
    [
        'month',
        'users_cnt',
        'sessions_cnt',
        'purchases_cnt',
        'conversion_pct',
        'avg_purchase_price',
        'median_purchase_price',
        'avg_tickets_num',
        'median_tickets_num'
    ]
]

monthly_summary

,month,users_cnt,sessions_cnt,purchases_cnt,conversion_pct,avg_purchase_price,median_purchase_price,avg_tickets_num,median_tickets_num
0,2025-01-01,1053,4783,394,8.24,14490.024286,9842.340,1.477752,1.0
1,2025-02-01,1027,4002,263,6.57,13810.117085,8649.830,1.484746,1.0
2,2025-03-01,1118,4362,294,6.74,14153.207151,10141.205,1.523256,1.0
3,2025-04-01,1147,4471,280,6.26,13592.214188,10042.880,1.464286,1.0
4,2025-05-01,1204,4650,282,6.06,14945.061497,9200.960,1.611465,1.0
5,2025-06-01,1157,4575,264,5.77,15003.014877,10810.540,1.557895,1.0
6,2025-07-01,1178,4061,82,2.02,16754.383667,10849.350,1.677778,1.0
7,2025-08-01,1148,4388,293,6.68,17365.344455,12142.740,1.784848,1.0
8,2025-09-01,953,3231,210,6.50,17100.228106,12256.970,1.748899,1.0


### Сводка ключевых метрик по месяцам

Итоговая таблица подтверждает, что июль 2025 года является аномальным месяцем в исследуемом периоде. Несмотря на сопоставимый объём пользователей и сессий относительно соседних месяцев, число покупок в июле резко снижается до 82, а конверсия в покупку падает до 2.02%. Уже в августе конверсия восстанавливается до 6.68%, что дополнительно указывает на локальный характер июльской просадки.

При этом в июле и последующих месяцах наблюдается рост средней и медианной цены покупки, а также среднего количества билетов в заказе. Это означает, что аномалия июля связана не с уменьшением входящего трафика, а с резким ухудшением прохождения нижней части воронки. Рост чека и размера заказа может сопровождать изменения в структуре покупок, но сам по себе не объясняет июльскую просадку, поскольку в августе и сентябре эти показатели остаются высокими, а конверсия восстанавливается.

## 8. Основные выводы

1. **Главные потери воронки сосредоточены на переходе `Выдача → Оффер` и на нижних этапах оформления.**
   Переход `Выдача → Оффер` даёт крупнейшую абсолютную потерю пользователей, а шаги `Оффер → Чекаут` и `Чекаут → Успешная покупка` показывают самую низкую относительную конверсию.

2. **Этап выдачи — одна из ключевых точек для дополнительной продуктовой диагностики и оптимизации.**
   Дополнительный анализ показал, что сессии с быстрым переходом из выдачи в оффер чаще продолжают движение по воронке, тогда как при длительном времени на экране выдачи вероятность перехода снижается. Это может указывать на сложность выбора, недостаточную релевантность части предложений или перегруженность сценария выбора.

3. **В июле 2025 года наблюдается аномальная просадка конверсии в покупку.**
   При сопоставимом объёме пользователей и сессий число покупок и итоговая конверсия в июле резко снижаются, а сравнение воронки по месяцам показывает, что основное отставание начинается на нижних этапах.

4. **Рост цены и размера заказа сопровождает июльский период, но не объясняет просадку полностью.**
   Начиная с июля растут средняя цена покупки и среднее количество билетов, однако в августе и сентябре эти показатели остаются высокими, тогда как конверсия восстанавливается.

Таким образом, основными зонами для продуктовой оптимизации выглядят этап выдачи и нижняя часть оформления, а июльская аномалия требует дополнительной проверки на уровне качества офферов и процесса покупки.

## 9. Продуктовые гипотезы

На основе проведённого анализа я сформулировал 3 продуктовые гипотезы, направленные на улучшение ключевых этапов клиентской воронки.

### Гипотеза 1. Улучшение экрана выдачи может повысить переход в оффер
Переход `Выдача → Оффер` даёт крупнейшую абсолютную потерю пользователей в воронке. Кроме того, сессии с длительным временем на экране выдачи заметно реже доходят до оффера, чем сессии с более быстрым переходом.
Если на экране выдачи визуально выделять наиболее релевантные и выгодные предложения, то пользователям будет проще ориентироваться в результатах поиска, и конверсия `Выдача → Оффер` может вырасти.

**Продуктовое решение:**
Добавить персонализированную подборку, рекомендательный блок или визуально выделить наиболее релевантные и выгодные варианты в верхней части выдачи.

**Ожидаемый эффект:**
- рост CR `Выдача → Оффер`;
- снижение доли быстрых выходов с экрана выдачи;
- рост итоговой конверсии в покупку.

### Гипотеза 2. Возвратные коммуникации для пользователей, не завершивших покупку, повысят конверсию
Существенная часть пользователей доходит до этапов выдачи и оффера, но не завершает покупку в рамках текущей сессии.
Если запускать персонализированные пуши, письма или ретаргетинговые коммуникации для пользователей, которые дошли до выдачи или оффера, но не завершили покупку, то часть из них вернётся в воронку и завершит заказ.

**Продуктовое решение:**
Запустить персонализированные напоминания и ретаргетинговые коммуникации для пользователей, которые были на этапе выдачи или оффера, но не завершили покупку.

**Ожидаемый эффект:**
- рост доли возвратов в воронку;
- увеличение числа покупок без пропорционального роста затрат на привлечение.

### Гипотеза 3. Улучшение условий оффера в сезонный период может повысить конверсию в покупку
В июле 2025 года основная просадка наблюдается не в верхней части воронки, а начиная с этапа оффера и дальше — на шагах оформления и покупки.
Если в период сезонного спроса предлагать пользователям более конкурентные условия — например, специальные предложения, скидки на чувствительные сегменты или более заметную коммуникацию выгодных офферов, — то переход из оффера в чекаут и итоговая конверсия в покупку могут вырасти.

**Продуктовое решение:**
В сезонные периоды выделять наиболее выгодные предложения, тестировать специальные условия для чувствительных сегментов и усиливать коммуникацию ценности оффера на этапе выбора.

**Ожидаемый эффект:**
- рост CR `Оффер → Чекаут`;
- рост итоговой конверсии в покупку;
- снижение сезонной просадки конверсии.

### Наиболее перспективная гипотеза

Наиболее перспективной выглядит **гипотеза об улучшении экрана выдачи**.

Именно эта гипотеза выглядит наиболее перспективной, потому что она напрямую опирается на главный вывод исследования: крупнейшая абсолютная потеря пользователей происходит на переходе `Выдача → Оффер`. Кроме того, дополнительный анализ времени на выдаче показал, что длительное взаимодействие с этим этапом связано с более низкой вероятностью перехода дальше по воронке. Улучшение сценария выбора на выдаче потенциально способно повлиять сразу на большой объём трафика и повысить эффективность всей воронки, а не только её финального этапа.

**Как я бы проверял эту гипотезу:**
Я бы запустил A/B-тест, в котором пользователи случайным образом делятся на контрольную и тестовую группы. В контрольной группе экран выдачи остаётся без изменений, а в тестовой добавляется улучшенный сценарий — например, визуальное выделение лучших предложений, рекомендательный блок или более заметная подача выгодных вариантов.

Основной метрикой эксперимента я бы выбрал конверсию `Выдача → Оффер`, так как именно этот шаг напрямую отражает эффект изменения интерфейса. В качестве дополнительных метрик я бы отслеживал итоговую конверсию в покупку, долю быстрых выходов с экрана выдачи и возможные негативные эффекты, например увеличение времени выбора без роста переходов дальше по воронке.

Если в тестовой группе переход `Выдача → Оффер` и итоговая конверсия окажутся выше без ухудшения пользовательского поведения на следующих шагах, гипотезу можно будет считать подтверждённой.